File: src/main/java/dev/andante/bubble/command/BubbleCommand.java
package dev.andante.bubble.command;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import dev.andante.bubble.world.TemporaryWorld;
import dev.andante.bubble.BubbleManager;
import net.minecraft.command.argument.DimensionArgumentType;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.text.Text;

import static net.minecraft.server.command.CommandManager.*;

public class BubbleCommand {
    public static final DynamicCommandExceptionType INVALID_DIMENSION_EXCEPTION = new DynamicCommandExceptionType(id -> Text.translatable("argument.dimension.invalid", id));

    public static void register(CommandDispatcher<ServerCommandSource> dispatcher) {
        dispatcher.register(
                literal("bubble")
                        .requires(s -> s.hasPermissionLevel(2))
                        .then(
                                literal("add")
                                        .executes(BubbleCommand::executeAdd)
                        )
                        .then(
                                literal("delete")
                                        .then(
                                                argument("dimension", TemporaryDimensionArgumentType.dimension())
                                                        .executes(BubbleCommand::executeDelete)
                                        )
                        )
        );
    }

    private static int executeAdd(CommandContext<ServerCommandSource> context) {
        ServerCommandSource source = context.getSource();
        MinecraftServer server = source.getServer();
        BubbleManager bubbleManager = BubbleManager.getOrCreate(server);
        TemporaryWorld world = bubbleManager.createAndInitialize();
        source.sendFeedback(Text.literal("Successfully added dimension '%s'".formatted(world.getRegistryKey().getValue())), true);
        return 1;
    }

    public static int executeDelete(CommandContext<ServerCommandSource> context) throws CommandSyntaxException {
        ServerWorld world = DimensionArgumentType.getDimensionArgument(context, "dimension");
        if (world instanceof TemporaryWorld temporaryWorld) {
            ServerCommandSource source = context.getSource();
            MinecraftServer server = source.getServer();
            BubbleManager bubbleManager = BubbleManager.getOrCreate(server);
            bubbleManager.scheduleDelete(temporaryWorld);
            source.sendFeedback(Text.literal("Successfully deleted dimension '%s'".formatted(temporaryWorld.getRegistryKey().getValue())), true);
            return 1;
        } else {
            throw INVALID_DIMENSION_EXCEPTION.create(world.getRegistryKey().getValue());
        }
    }
}

File: src/main/java/dev/andante/bubble/command/BubbleCommand.kt
package dev.andante.bubble.command

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import dev.andante.bubble.Bubble
import dev.andante.bubble.BubbleManager
import dev.andante.bubble.world.BubbleWorld
import net.minecraft.command.argument.DimensionArgumentType
import net.minecraft.server.command.CommandManager
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.text.Text

object BubbleCommand {
    private val INVALID_DIMENSION_EXCEPTION = DynamicCommandExceptionType { id -> Text.translatable("argument.dimension.invalid", id) }

    fun register(dispatcher: CommandDispatcher<ServerCommandSource>) {
        dispatcher.register(
            CommandManager.literal("bubble-test")
                .requires { source -> source.hasPermissionLevel(2) }
                .then(
                    CommandManager.literal("add")
                        .executes(::executeAdd)
                )
                .then(
                    CommandManager.literal("delete")
                        .then(
                            CommandManager.argument("dimension", DimensionArgumentType.dimension())
                                .suggests(Bubble::suggestBubbleDimensions)
                                .executes(::executeDelete)
                        )
                )
        )
    }

    private fun executeAdd(context: CommandContext<ServerCommandSource>): Int {
        val source = context.source
        val server = source.server
        val bubbleManager = BubbleManager.getOrCreate(server)
        val world = bubbleManager.createAndInitialize()
        source.sendFeedback({ Text.literal("Successfully added dimension '${world.worldRegistryKey.value}'") }, true)
        return 1
    }

    @Throws(CommandSyntaxException::class)
    fun executeDelete(context: CommandContext<ServerCommandSource>): Int {
        val world = DimensionArgumentType.getDimensionArgument(context, "dimension")
        return if (world is BubbleWorld) {
            val source = context.source
            val server = source.server
            val bubbleManager = BubbleManager.getOrCreate(server)
            bubbleManager.remove(world)
            source.sendFeedback({ Text.literal("Successfully deleted dimension '${world.worldRegistryKey.value}'")}, true)
            1
        } else {
            throw INVALID_DIMENSION_EXCEPTION.create(world.registryKey.value)
        }
    }
}
